MINI-PROJECTS.md

```diff
@@ -0,0 +1,103 @@
+CBMC Mini Projects
+==================
+
+The following projects are short, focussed features that give new CBMC
+contributors an introduction to one part of the codebase. If you're
+interested in contributing to CBMC, feel free to start with any of these
+projects!
+
+
+Task List
+---------
+
+### `__CPROVER_print`
+
+**Task**: Implement a CPROVER intrinsic that assigns an expression to a
+dummy value, to help with debugging and tracing.
+
+**Background**:
+Inserting `printf("value of expr: %d\n", expr);` is a common debugging
+and tracing technique. The CBMC equivalent is to assign an expression to
+a temporary variable and run `cbmc --trace`:
+
+```c
+$ cat foo.c
+#define __CPROVER_print(var) { int value_of_##var = (int) var; }
+
+void foo(int x)
+{
+  __CPROVER_print(x);
+  assert(0);  // Make CBMC halt its exploration and dump a trace up to
+              // this point
+}
+
+int main()
+{
+  foo(3);
+}
+```
+
+This has several limitations (we need different variants of the macro
+for pointers, etc). It would be great to have `__CPROVER_print` as a
+CPROVER intrinsic, so that users don't need to do the awkward define or
+manual ghost assignment in their code.
+
+**Hints**: Figure out how `__CPROVER_assume` and `__CPROVER_assert` work.
+Then figure out how to add new instructions into a goto-program.
+
+
+### Annotation to constrain function pointers
+
+**Task**: implement a CPROVER intrinsic that tells CBMC what function a
+function pointer points to.
+
+**Background**: CBMC often doesn't know which function a function
+pointer points to. If a function is called through a function pointer,
+CBMC currently assumes that *any* function that has the same signature
+could be called. This results in a combinatorial explosion of paths when
+CBMC is exploring the program. It would be useful to have an annotation
+that tells CBMC exactly what function a pointer is expected to point to.
+
+
+### Generalize loop-unwinding bounds to cover multiple loops
```

Comment:
Can we add a bit more information here to say that it should be done in cbmc and in the goto symex part? It's probably a bit hard to figure out where to start from the current description.

Reply:
Thanks, I've added a few hopefully helpful hints.